Purchases/Networking/Caching/CallbackCache.swift

```diff
@@ -0,0 +1,55 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  CallbackCache.swift
+//
+//  Created by Joshua Liebowitz on 11/18/21.
+
+import Foundation
+
+class CallbackCache<T> where T: CachableCallback {
+
+    var cache: [String: [T]] = [:]
+    let callbackQueue: DispatchQueue
+
+    init(callbackQueue: DispatchQueue) {
+        self.callbackQueue = callbackQueue
+    }
+
+    func add(callback: T) -> CallbackCacheStatus {
+        callbackQueue.sync {
+            var values = cache[callback.key] ?? []
+            let cacheStatus: CallbackCacheStatus = !values.isEmpty ?
+                .addedToExistingInFlightList :
+                .firstCallbackAddedToList
+
+            values.append(callback)
```

Comment:
i might answer my own question as i continue reviewing, but should we be appending `callback.callback`? (and probably renaming if so?)

right now it looks like the cache is essentially [key : [[key: value]] (though i know that inner key : value is an object, not a dictionary...)

Reply:
I wish that could have been the case, but when it comes to generics, you can't use blocks since blocks aren't a concrete type.
I want to store a block that can have any params, so it would have been nice if we could say "I want this to be a generic block" - nope. We need a specific type for the `T` in `var cachedCallbacksByKey: [String: [T]] = [:]`. 
That's why I create the protocol `CachableCallback` and restricted the `CallbackCache` to only work for types that conform to it. Whatever class implements the`CachableCallback` stores the block we want to cache. Kinda like how you need to do some boxing/unboxing in Java.